### PR TITLE
Only persist trip with transit and no-rental itinerary

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [X] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [X] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds a check to `MonitoredTripController` so that only trips with transit and no rentals can be saved/monitored.

To test in OTP-RR, use any branch except https://github.com/opentripplanner/otp-react-redux/pull/266.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package org.opentripplanner.middleware.utils;
+
+import org.opentripplanner.middleware.otp.response.Itinerary;
+import org.opentripplanner.middleware.otp.response.Leg;
+
+import java.util.Collection;
+import java.util.List;
+
+public class ItineraryUtils {
+    private static final Collection<String> OTP_TRANSIT_MODES = List.of("TRAM", "BUS", "SUBWAY", "FERRY", "RAIL", "GONDOLA");
```
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: These possible modes are missing: `CABLE_CAR`, `FUNICULAR`, `AIRPLANE`

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    public static boolean itineraryHasRental(Itinerary itinerary) {
+        if (itinerary != null && itinerary.legs != null) {
+            for (Leg leg : itinerary.legs) {
+                if (leg.mode.contains("_RENT")) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
```
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: These 3 methods could be combined into one where each leg is checked for all info. Also, this is missing a check for a ridehail.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.opentripplanner.middleware.otp.response.Leg;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+public class ItineraryUtilsTest {
+    @ParameterizedTest
+    @MethodSource("createItineraryHasTransitNoRentalsCases")
+    public void testItineraryHasTransitNoRentals(Itinerary itinerary, boolean expectedResult) {
+        Assertions.assertEquals(expectedResult, ItineraryUtils.itineraryHasTransitAndNoRentals(itinerary));
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: It's much nicer to include a test case message so that it's easier to see in the test case output which test has failed if one does.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        Itinerary itineraryWithTransitAndRental = new Itinerary();
+        itineraryWithTransitAndRental.legs = List.of(walkLeg, transitLeg, rentalLeg);
+
+        return Stream.of(
+            Arguments.of(itineraryWithTransitNoRentals, true),
+            Arguments.of(itineraryWithoutTransit, false),
+            Arguments.of(itineraryWithTransitAndRental, false),
+            Arguments.of(blankItinerary, false),
+            Arguments.of(null, false)
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Needs a test case for ridehailing.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comments

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: We'll have to add them to OTP-UI too!

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/8cc7dd929604fe9ba216dcfe41d5ed59161da22c.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java (hunk 3 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/8cc7dd929604fe9ba216dcfe41d5ed59161da22c.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java (hunk 2 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/8cc7dd929604fe9ba216dcfe41d5ed59161da22c.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: In https://github.com/ibi-group/otp-middleware/pull/91/commits/8cc7dd929604fe9ba216dcfe41d5ed59161da22c, I ended up using fields `leg.transitLeg`, `leg.rentedVehicle` and `leg.hailedCar` that are populated by OTP.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        boolean hasTransit = false;
+        boolean hasRentalOrRideHail = false;
+
+        if (itinerary != null && itinerary.legs != null) {
+            for (Leg leg : itinerary.legs) {
+                if (leg.transitLeg != null && leg.transitLeg) {
+                    hasTransit = true;
+                }
+                if (leg.rentedVehicle != null && leg.rentedVehicle ||
+                    leg.hailedCar != null && leg.hailedCar) {
```
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: This is missing a check for `leg.rentedBike` and `leg.rentedCar`.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private static Stream<Arguments> createItineraryCanBeMonitoredCases() {
+        Leg transitLeg = new Leg();
+        transitLeg.mode = "BUS";
+        transitLeg.transitLeg = true;
+
+        Leg rentalLeg = new Leg();
+        rentalLeg.mode = "BICYCLE_RENT";
+        rentalLeg.rentedVehicle = true;
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: `rentedVehicle` is only set to true if the mode is `MICROMOBILITY_RENT`.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        Itinerary itineraryWithTransitAndRental = new Itinerary();
+        itineraryWithTransitAndRental.legs = List.of(walkLeg, transitLeg, rentalLeg);
+
+        Itinerary itineraryWithTransitAndRideHail = new Itinerary();
+        itineraryWithTransitAndRideHail.legs = List.of(walkLeg, transitLeg, rideHailLeg);
+
+        return Stream.of(
+            Arguments.of(itineraryWithTransitNoRentals, true, "Itinerary with transit, no rentals/ride hail."),
+            Arguments.of(itineraryWithoutTransit, false, "Itinerary without transit."),
+            Arguments.of(itineraryWithTransitAndRental, false, "Itinerary with transit and rental."),
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Please differentiate between `bikeRental`, `carRental` and `vehicleRental`.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comments.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package org.opentripplanner.middleware.utils;
+
+import org.opentripplanner.middleware.otp.response.Itinerary;
+import org.opentripplanner.middleware.otp.response.Leg;
+
+public class ItineraryUtils {
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: Should the class have JavaDoc?

**Review by br648 — APPROVED** [review]
One minor comment, but looks good to me. Approving pending resolution of Evan's comments.

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java [hunk 9]
```diff
@@ -26,21 +27,29 @@ public MonitoredTripController(String apiPrefix) {
...
-        } catch (Exception e) {
-            logMessageAndHalt(
-                req,
-                HttpStatus.BAD_REQUEST_400,
-                "Invalid input data received for monitored trip.",
-                e
-            );
-        }
+        checkItineraryCanBeMonitored(monitoredTrip, req);
+        processTripQueryParams(monitoredTrip, req);
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:59]: Merge conflict alert (with #77)! Probably unavoidable 🤷‍♂️

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java [hunk 10]
```diff
@@ -75,4 +77,18 @@ private void verifyBelowMaxNumTrips(String userId, Request request) {
             );
         }
     }
+
+    /**
+     * Checks that the given {@link MonitoredTrip} can be monitored
+     * (the underlying {@link org.opentripplanner.middleware.otp.response.Itinerary} has transit and no rentals).
+     */
+    private void checkItineraryCanBeMonitored(MonitoredTrip trip, Request request) {
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java]: Might be better to just accept the Itinerary as an arg (instead of MonitoredTrip). Better name might be `validateItinerary`.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java]: ... except that there is no real validation or issue reporting performed.

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java [hunk 11]
```diff
@@ -75,4 +77,18 @@ private void verifyBelowMaxNumTrips(String userId, Request request) {
...
+        if (!ItineraryUtils.itineraryCanBeMonitored(trip.itinerary)) {
+            logMessageAndHalt(
+                request,
+                HttpStatus.BAD_REQUEST_400,
+                "Only trips with an itinerary that includes transit and no rentals or ride hailing can be monitored."
+            );
+        }
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: Instead of embedding the logic deep within ItineraryUtils, it might be better to have independent methods be invoked here so the error message can be tailored a little more easily and the logic is a bit more transparent. For example:

```java
      if (!ItineraryUtils.itineraryHasTransit(itinerary)) {
            logMessageAndHalt(
                request,
                HttpStatus.BAD_REQUEST_400,
                "Monitored trip must reference an itinerary that includes transit legs."
            );
        }
       if (ItineraryUtils.itineraryHasRental(itinerary)) {
            logMessageAndHalt(
                request,
                HttpStatus.BAD_REQUEST_400,
                "Monitored trip must not reference an itinerary that includes legs using rental vehicles."
            );
        }
```

I realize that this might be at odds with a suggestion that @evansiroky made in the UI PR about checking the legs in the same loop. But, if we're being real, running through two loops of an itinerary's legs instead of one is a negligible cost (and I think readability of code is a higher priority).

**Review by landonreed — CHANGES_REQUESTED** [review]
See comments about splitting/surfacing the validation checks.

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java (hunk 11 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: Who should break the tie?

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: If a trip cannot be monitored, I think there should be a single message saying all the conditions that should be met. ~That reminds me to update the swagger method description too...~

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java (hunk 8 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: Added in https://github.com/ibi-group/otp-middleware/pull/91/commits/43425abecb53ee8d559c0d7c75062e6c84b7f03d.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java (hunk 7 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/43425abecb53ee8d559c0d7c75062e6c84b7f03d.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java (hunk 6 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Thanks for the clarification. Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/43425abecb53ee8d559c0d7c75062e6c84b7f03d.

File: src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java (hunk 5 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/43425abecb53ee8d559c0d7c75062e6c84b7f03d.

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java (hunk 10 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java]: Renamed to `checkTripCanBeMonitored` in https://github.com/ibi-group/otp-middleware/pull/91/commits/2c07661450ae3c3531b35660a0892ab10bb6d534...

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java (hunk 11 above)
Comment (evansiroky) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: I like the idea of providing a list of criteria that must be met.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        itineraryWithRentalBikeWithoutTransit.legs = List.of(walkLeg, rentalBikeLeg);
+
+        Itinerary itineraryWithTransitAndRentalBike = new Itinerary();
+        itineraryWithTransitAndRentalBike.legs = List.of(walkLeg, transitLeg, rentalBikeLeg);
+
+        Itinerary itineraryWithTransitAndRentalCar = new Itinerary();
+        itineraryWithTransitAndRentalCar.legs = List.of(walkLeg, transitLeg, rentalCarLeg);
+
+        Itinerary itineraryWithTransitAndRentalMicromobility = new Itinerary();
+        itineraryWithTransitAndRentalMicromobility.legs = List.of(walkLeg, transitLeg, rentalMicromobilityLeg);
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: Doesn't seem to be added to list of test cases.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+        Itinerary itineraryWithTransitAndRideHail = new Itinerary();
+        itineraryWithTransitAndRideHail.legs = List.of(walkLeg, transitLeg, rideHailLeg);
+
+        return Stream.of(
```
Comment (evansiroky) [inline src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: This does seem simpler than creating a test case class, but it might result in runtime errors if an incorrect number or type of arguments are entered. I'm fine with keeping this as is for now and it's good to know this is a possible way of creating arguments for a parameterized test.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See https://github.com/ibi-group/otp-middleware/pull/91#discussion_r516970277 and https://github.com/ibi-group/otp-middleware/pull/91#discussion_r516970742.

File: src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java (hunk 12 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/utils/ItineraryUtilsTest.java]: So many things to forget...

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java (hunk 11 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: @evansiroky please clarify what needs to change, thank you.

Comment (landonreed) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: I think what @evansiroky is suggesting is we collect a list of validation issues. So, it would be similar to the code I posted above, but the halt should only happen at the end of the method and it would include the full list of validation issues.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:178]: Let's try this: https://github.com/ibi-group/otp-middleware/pull/91/commits/83405f10b950ed5b4fb6ca75ddfa1a1f34753d1e.

**Review by evansiroky — APPROVED** [review]

File: src/test/java/org/opentripplanner/middleware/models/ItineraryTest.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.opentripplanner.middleware.otp.response.Itinerary;
+import org.opentripplanner.middleware.otp.response.Leg;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+public class ItineraryTest {
```
Comment (landonreed) [inline src/test/java/org/opentripplanner/middleware/models/ItineraryTest.java:17]: Please add javadoc.

File: src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java [hunk 15]
```diff
@@ -79,6 +79,39 @@
...
+                if (leg.transitLeg != null && leg.transitLeg) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether the itinerary includes transit.
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java]: copy/pasta 🍝

File: src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java [hunk 16]
```diff
@@ -79,6 +79,39 @@
...
+     * @return true if at least one {@link Leg} of the itinerary is a rental or ride hail leg per OTP.
+     */
+    public boolean hasRentalOrRideHail() {
+        if (legs != null) {
+            for (Leg leg : legs) {
+                if (leg.rentedBike != null && leg.rentedBike ||
+                    leg.rentedCar != null && leg.rentedCar ||
+                    leg.rentedVehicle != null && leg.rentedVehicle ||
+                    leg.hailedCar != null && leg.hailedCar) {
```
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java]: Here and in the hasTransit method, I'm not crazy about all of the `null` checks.

It looks like this is an alternative approach: `Boolean.TRUE.equals(leg.rentedBike)`. I'm somewhat in favor of this approach because it's a bit shorter and provides a more easily replicable approach for other cases we might have in the future.

Also, probably good to add a test case or two with null values to ensure we don't throw an NPE.

Ref: https://stackoverflow.com/a/18790171/915811

**Review by landonreed — CHANGES_REQUESTED** [review]
Just a couple of minor things, otherwise looks good!

File: src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java (hunk 16 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java]: I like that better too, thanks for pointing out (see https://github.com/ibi-group/otp-middleware/pull/91/commits/134a1c76a195169fd9c0a2c196409b3b4664f57c). There are already some cases with null values in the tests (anywhere the `rented*/hailed` field is not set).

File: src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java (hunk 15 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java]: I was probably hungry

File: src/test/java/org/opentripplanner/middleware/models/ItineraryTest.java (hunk 14 above)
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/models/ItineraryTest.java:17]: Updated in https://github.com/ibi-group/otp-middleware/pull/91/commits/134a1c76a195169fd9c0a2c196409b3b4664f57c

**Review by landonreed — APPROVED** [review]
